### ISSUE TITLE
terminal activity indicator/alert when in background

### ISSUE THREAD

**Issue by damnskippy** [microsoft/terminal#7955]
Problem:
Often times there's a command invoked on the shell prompt that's long running and takes a while, and the user switches to a different window and puts the terminal window in the background while waiting for this command/job to complete.  When the job completes, user is unaware this happened until the terminal window is switched back to (foreground) or user periodically peeks to see if the job has completed.

Suggestion:
If there can be some some sort of visual indicator or alert when there is terminal activity to suggest there is new output logged into the terminal, it would provide a cue to the user that the job may be complete.  If the long running job happens to continually output to the terminal this wouldn't be as useful, but most times there's a job that runs for an unpredictable amount of time and then terminal returns to the prompt.

The indicator could be a temporary flashing of the window title in the task bar, and/or a change to a different shade/color until user has acknowledged the alert.  Or may be display a badge on the title bar.  

Alternatives considered that are not viable or great:
- use unix terminal bell (I couldn't get this to work reliably while ssh'ed into different servers, also this requires use have the sound/headphones on, and NOT disable the bell on the terminal).
- use some type of window pop up (would require X or some such windowing to accomplish this, and not native to terminal).
- notification systems similar to growl/notifier/etc (in MacOS) or even the Windows notification are not accessible while ssh'ed into a remote server where such long running command is run.

**Comment by damnskippy** [comment]
@j4james @zadjii-msft,  
Here I suggested a terminal output/activity based alert mechanism.  This detection will be so native to the terminal, the user doesn't have to explicitly say "notify me when done" or in cases when the user may not have had a chance to say "notify me" (user didn't anticipate/predict the command to take long to finish), and yet wants to see a notification.
Using terminal output activity as the trigger will mean it will work on both local and remote (ssh) sessions.
#7718 while similar seems to be a slightly different use case where the user explicitly has to request to be notified via the OSC.  I think both use cases have their place.  The output based notification falls flat when there the long running command is continually spewing output.
Thanks.

**Comment by damnskippy** [comment]
I don't use a mac, but it seems iterm has a feature like this based on this information in this [post](https://apple.stackexchange.com/a/22354).  I've seen it in one other terminal emulator as well (SecureCRT I think).

**Comment by zadjii-msft** [comment]
So, if this was to be implemented, and someone were to run something like a build that was displaying output constantly, wouldn't the indicator just be perpetually on? I don't really see the value in that, so there would have to be some sort of contextual way to enable/disable this. 

Though I guess reading that post, the indeterminate progress wheel as the indicator kinda makes sense - then the wheel _stopping_ spinning would be the indicator that the process stopped running. My only great disappointment is there's not a good way for the terminal to display that indicator in the taskbar icon. We could easily add the indicator to the tab itself, but that would only be visible if the window was visible. (see also #3991)

I personally think that this is better served by a combo of #3004 and having the client application emit the corresponding sequence. 

Maybe we could combine the activity spinner with the indeterminate taskbar state. Add a setting to allow the user to show indeterminate state in the taskbar while the indicator is active. We'd probably want that to be overridden by whatever the client sets from the VT sequences as described in #3004, but that's not out of the realm of possibility.

**Comment by damnskippy** [comment]
Impressive how you are able to keep these "related" issues in the head and can call out at will and tie them together :-)
It's a very useful feature IMHO (shamelessly plugging my request again).

In terms of usage, one could slice the end user requirement/experience in a couple of ways:
- a long running command that doesn't continually spew output (such as a build), versus, commands that run for a while without outputting anything but when they end, terminal returns to the prompt.  And in both cases having the terminal activity indicated to the user in some form is useful depending on the use case, the latter being more useful scenario.
- user knows ahead of time and explicitly requests an alert, to be notified, at the end of the long running command, versus running a command that the user is oblivious to how long the command will take to complete, or didn't expect for it to take long when launching the command unsuspectingly but wants to be notified of the status in the background.

It seems the UI options you've suggested above covers both cases - please correct me if I'm wrong.

Thanks as always.

**Comment by KalleOlaviNiemitalo** [comment]
Both tmux and GNU Screen have a "silence" feature that notifies the user when there has been no new output for some time. In GNU Screen, the default is 30 seconds if enabled.

**Comment by naikel** [comment]
Most of the terminal emulator apps issue some kind of indicator when the buffer changes and it's not in the active view.  For example you can be running a long task in one tab and have another active tab and there should be an indication that there was a buffer change in any other tab.  Like the tab title change from black to red foreground (red font).  That's extremely important.

**Comment by naikel** [comment]
I know probably this is not what you guys wanted but here's my version of something similar if a non-focused tab has new output:

![naikelGintoki-2021-01-09-11-09-1](https://user-images.githubusercontent.com/7663686/104106740-65fadf00-5285-11eb-805e-5770a380c618.gif)

**Comment by skyline75489** [comment]
I think this is a reasonable feature request. AFAIK, iTerm2 has this and I think it's useful. The taskbar indicator (OSC 9;4) is useful, but not suitable for multiple tabs.

> someone were to run something like a build that was displaying output constantly, wouldn't the indicator just be perpetually on?

Well yeah. But this does not bother me.

**Comment by iuricarraro** [comment]
MobaXterm has similar feature and very useful.
![1](https://user-images.githubusercontent.com/4460055/112541469-2f1b6c00-8d92-11eb-8611-e1e6317b6b47.png)

**Comment by zadjii-msft** [comment]
As an example from the wild: Gnome's Console will actually [send a desktop notification when the commandline finishes](https://itsfoss.com/gnome-console/), if it's in the background. I don't think we could possibly get the actual commandline without much deeper shell integration (or heuristics based on typing enter, which I don't love).

See also:
* #7718
* https://gitlab.freedesktop.org/terminal-wg/specifications/-/issues/13

**Comment by warpdesign** [comment]
How about something like this?
![image](https://user-images.githubusercontent.com/199648/194486700-11fd08a1-a11f-4059-b760-e83b09d198fd.png)

I also like @naikel's suggestion: I guess this could be configurable.

**Comment by max-programming** [comment]
Is this being worked on? Or is there any plugin or alternative to achieve this?

**Comment by zadjii-msft** [comment]
> Is this being worked on? 

Nope, not yet. We haven't really settled on a design for this feature yet. Do you have a particular design in mind?

Rereading this thread, I think I like the design in https://github.com/microsoft/terminal/issues/7955#issuecomment-1271186689 the most myself. Kinda like the tab bell notification we have currently when an inactive tab BELs, but, just for _any output_.

**Comment by damnskippy** [comment]
> notification: send a desktop notification if an inactive tab recieves output?

Is this a toast notification or some thing like that?  If yes, perhaps it can show the window title and any other useful info in it, and when clicked on the notification could switch to the window (raise to foreground) that the notification is for?
Thanks.

**Comment by zadjii-msft** [comment]
note to self: an example of the `notification` style:

https://user-images.githubusercontent.com/18356694/227358435-8c4e7533-e4a9-4740-9779-0e29c9b1d195.mp4



credit:
https://mastodon.social/@raggesilver/110045050437013369

**Comment by naikel** [comment]
@maddishah at the time I modified the source code of this project for that, but I never send a pull request for it because it seemed nobody liked the design for it.

It's long gone now.

**Comment by fdcastel** [comment]
Just passing to say that I've been working with computers for over 30 years. 

And, throughout that time, I've often dreamed of having something like this.

_Especially... right now, as I wait for a long build task to finish._ 

I'm eager to see this one become a reality!

**Comment by xgdgsc** [comment]
I also need feature like https://unix.stackexchange.com/questions/617889/is-there-a-terminal-emulator-which-can-notify-when-there-is-an-output-with-certa .

**Comment by zadjii-msft** [comment]
Best way to show your support for this issue is by 👍'ing it - that's generally how we track priority around these parts.

(I also want you to do that because **I** want this feature too so I want all the excuses possible to work on it 😆)

**Comment by etxaleku** [comment]
I would like this feature as well please.  👍

**Comment by kimsey0** [comment]
If this flashing/color-change state can be triggered programmatically (e.g., via a terminal bell or a dedicated escape sequence), it would also be useful for AI coding agents that need to request user attention.

**Comment by heaths** [comment]
I've been using Ghostty in macOS that does this with OSC 133;C and OSC 133;D in bash and other supported shells. I'm working on pwsh support now.

Basically, your pre-prompt (`$PS0` in bash, `$function:PSConsoleHostReadLine` in pwsh) needs to send OSC 133;C e.g.,

```pwsh
"`e]133;C;cmdline_url=$([Uri]::EncodeUriString($the_command))`a" # cmdline_url is the standard and optional, but "cmdline" variations are common
```

Your prompt (`$PS1` in bash, `$function:prompt` in pwsh) needs to send OSC 133;D e.g.,

```pwsh
"`e]133;D;$LASTEXITCODE`a" # optional exit code, though you'd want to capture it probably because it often gets overwritten
```

You measure the time between OSC 133;C and OSC 133;D and, when above some user-defined threshold (perhaps profile-based, so a default could be specified but overridden per-profile) and show a system notification when that threshold is crossed with the URI-decoded `cmdline_url`.
